sql/src/main/scala/com/github/sadikovi/spark/riff/DefaultSource.scala

```diff
@@ -212,10 +213,26 @@ class DefaultSource
       val reader = Riff.reader(hadoopConf, path)
       val iter = reader.prepareRead(predicate)
       Option(TaskContext.get()).foreach(_.addTaskCompletionListener(_ => iter.close()))
-      // TODO: this is inefficient, it would be better to apply projection directly on indexed row
+
       val td = reader.getFileHeader().getTypeDescription()
-      if (projectionFields.length < td.size()) {
+      if (metadataCountEnabled && projectionFields.isEmpty) {
+        // only perform optimization if it is enabled
+        // TODO: Move it into Riff format, once projection is fixed
+        var numRecords = reader.getFileHeader().getNumRecords()
+
+        new Iterator[InternalRow] {
+          override def hasNext: Boolean = {
+            numRecords > 0
+          }
+
+          override def next: InternalRow = {
+            numRecords -= 1
+            new ProjectionRow(0)
```

Comment:
I think it is worth returning empty row instead of creating projection row with empty array.